### DISCUSSION TITLE
3D-LDM shows no progress in sampled images

### DISCUSSION THREAD

**Discussion by kapsner** [Project-MONAI/GenerativeModels#458] [Q&A] [answered]
Hi there,

I am trying to adapt the 3D LDM tutorial to our data using the [pytorch-lightning](https://lightning.ai/docs/pytorch/stable/) framework and I am struggling to get it running properly.

My problem is that while the training / validation loss curves are looking very fine, the sampled synthetic images I am logging to tensorboard each 10 epochs do show no progress at all.

I have no clue, if I have made so mistake in transposing the code to `lightning` or if I have maybe falsely parameterized the UNet (using medical gray-scale images, one channel, of size [B,1,32,96,96]).

Any help is very appreciated!! 

Currently, the implementation is as follows:

```python
# Tutorials:
# https://github.com/Project-MONAI/GenerativeModels/blob/main/tutorials/generative/3d_ldm/3d_ldm_tutorial.py
# https://lightning.ai/docs/pytorch/stable/notebooks/course_UvA-DL/08-deep-autoencoders.html

import lightning as L
import numpy as np
import torch
import torch.nn.functional as F
from monai.visualize import plot_2d_or_3d_image
from torch.nn import L1Loss

from torch.cuda.amp import GradScaler, autocast

from generative.inferers import LatentDiffusionInferer
from generative.networks.nets import DiffusionModelUNet
from generative.networks.schedulers import DDPMScheduler
import pandas as pd
import logging

from backend.datamodule import DataModule

def _compute_scale_factor(metadata: pd.DataFrame, dataset_dir: str, ae_model, batch_size: int = 8, use_persistent_cache: bool = False, device: str = "cuda:0"):

    # define scaling factor (https://github.com/Project-MONAI/GenerativeModels/blob/main/tutorials/generative/3d_ldm/3d_ldm_tutorial.py#L314)
    dm = DataModule(
        metadata=metadata,
        dataset_dir=dataset_dir,
        batch_size=batch_size,
        use_persistent_cache=use_persistent_cache
    )
    dm.setup()
    
    batch_sizes = []
    scale_factors = []

    # generate iterator
    for batch in dm.train_dataloader():
        batch_sizes.append(len(batch["image"]))
        with torch.no_grad():
            with autocast(enabled=True):
                z = ae_model.autoencoder.encode_stage_2_inputs(batch["image"].to(device))
                scale_factors.append(1/torch.std(z).item())

    # calc mean
    mean_scale_factor = np.sum(np.array(scale_factors) * np.array(batch_sizes)) / np.sum(np.array(batch_sizes))

    msg = f"Scaling factor set to {mean_scale_factor}"
    print(msg)
    logging.info(msg)
    return mean_scale_factor


class LDMUNet(L.LightningModule):
    def __init__(self, **kwargs):
        super().__init__()

        # log hyperparameters
        self.save_hyperparameters(ignore=['ae_model'])
        self.automatic_optimization = True

        self.training_step_outputs = []
        self.validation_step_outputs = []
        self.test_step_outputs = []

        self.autoencoder = kwargs["ae_model"].autoencoder
        self.autoencoder.requires_grad_(False)
        self.z_example = None

        # only log images from first batch of every tenth val-epoch
        self.logged_valid_imgs = False

        # finally, define network
        self.unet = DiffusionModelUNet(
            spatial_dims=3,
            in_channels=3,
            out_channels=3,
            num_res_blocks=2,
            num_channels=(64, 128, 256, 512),
            attention_levels=(False, False, True, True),
            norm_num_groups=32,
            norm_eps=1e-6,
            resblock_updown=False,
            num_head_channels=8,
            #num_head_channels=(0, 0, 512, 512),
            with_conditioning=False,
            transformer_num_layers=1,
            cross_attention_dim=None,
            num_class_embeds=None,
            upcast_attention=False,
            use_flash_attention=False
        )
        self.scheduler = DDPMScheduler(
            num_train_timesteps=1000,
            schedule="scaled_linear_beta",
            beta_start=0.0015,
            beta_end=0.0205
        )

        self.inferer = LatentDiffusionInferer(
            scheduler=self.scheduler,
            scale_factor=self.hparams.scale_factor,
            #sldm_latent_shape=[32, 12, 48, 48] # monai-generative from git: 894f2ec3452d1bedbcd72e9842e1fa1fb4535ce9
        )

        self.l1_loss = L1Loss()

    def forward(self, x):
        pass
    
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(
            params=self.unet.parameters(),
            lr=self.hparams.learning_rate,
            weight_decay=self.hparams.weight_decay
        )
        return optimizer

    def training_step(self, batch, batch_idx):
        step_results = self._shared_step(batch=batch, batch_idx=batch_idx, prefix="train")
        self.log(
            name="train_loss",
            value=step_results["loss"],
            prog_bar=True,
            logger=False,
            on_step=True,
            on_epoch=True,
            sync_dist=self.hparams.sync_dist,
            batch_size=step_results["batch_size"]
        )
        self.training_step_outputs.append(step_results)
        return step_results

    def on_train_epoch_end(self):
        self._shared_epoch_end(self.training_step_outputs, "train")
        self.training_step_outputs.clear()    

    def validation_step(self, batch, batch_idx):
        step_results = self._shared_step(batch=batch, batch_idx=batch_idx, prefix="valid")
        self.log(
            name="valid_loss",
            value=step_results["loss"],
            prog_bar=True,
            logger=False,
            on_step=True,
            on_epoch=True,
            sync_dist=self.hparams.sync_dist,
            batch_size=step_results["batch_size"]
        )
        self.validation_step_outputs.append(step_results)
        return step_results

    def on_validation_epoch_end(self):
        self._shared_epoch_end(self.validation_step_outputs, "valid")
        # reset logging for next validation step
        self.logged_valid_imgs = False
        self.validation_step_outputs.clear()
    
    def _shared_step(self, batch, batch_idx, prefix):
        x = batch["image"]
        ids = batch["name"]

        if self.z_example is None:
            self.z_example = self.autoencoder.encode_stage_2_inputs(x)
            print(f"Size of latent space images: {tuple(self.z_example.size())[1:]}")

        loss_dict = self._shared_ldm(
            images=x,
            ids = ids,
            prefix=prefix
        )
        return dict(loss_dict, **{"batch_size": len(x), "device": x.device})
    
    def _shared_ldm(self, images, ids, prefix):

        # Generate random noise
        # noise = torch.randn_like(self.z_example)
        # --> this does not work, if batch-size is different
        input_size = ((len(images), ) + tuple(self.z_example.size())[1:])
        noise = torch.rand(
            input_size,
            dtype=self.z_example.dtype,
            layout=self.z_example.layout,
            device=images.device
        )

        # Create timesteps
        timesteps = torch.randint(
            0, self.inferer.scheduler.num_train_timesteps, (images.shape[0],), device=images.device
        ).long()

        # Get model prediction
        noise_pred = self.inferer(
            inputs=images,
            autoencoder_model=self.autoencoder,
            diffusion_model=self.unet,
            noise=noise,
            timesteps=timesteps
        )

        #loss = F.mse_loss(noise_pred.float(), noise.float())
        loss = self.l1_loss(input=noise_pred.float(), target=noise.float())

        if prefix == "valid" and not self.logged_valid_imgs:
            if self.current_epoch % 10 == 0 and str(images.device) == self.hparams.log_device:
                noise = torch.randn(((8, ) + tuple(self.z_example.size())[1:]))
                noise = noise.to(images.device)
                self.scheduler.set_timesteps(num_inference_steps=1000)
                
                synthetic_images = self.inferer.sample(
                    input_noise=noise,
                    autoencoder_model=self.autoencoder,
                    diffusion_model=self.unet,
                    scheduler=self.scheduler
                )

                synth_images = synthetic_images.detach().cpu().numpy()

                for _i in range(0, synth_images.shape[0]):
                    plot_2d_or_3d_image(
                        data=synth_images,
                        step=self.global_step,
                        writer=self.logger.experiment,
                        index=_i,
                        max_channels=1,
                        frame_dim=-3,
                        tag="synthetic_img/" + str(_i + 1)
                    )
                self.logged_valid_imgs = True
        
        return {"loss": loss}
    
    def _shared_epoch_end(self, outputs, prefix):
        # concat batch sizes
        batch_sizes = np.stack(
            [x["batch_size"] for x in outputs]
        )

        # concat losses
        losses = np.stack(
            [x["loss"].item() for x in outputs]
        )

        # calculating weighted mean loss
        avg_loss = np.sum(losses * batch_sizes) / np.sum(batch_sizes)

        self.log(
            name="loss/" + prefix,
            value=avg_loss,
            prog_bar=False,
            logger=True,
            on_step=False,
            on_epoch=True,
            sync_dist=self.hparams.sync_dist
        )

```

**Comment by marksgraham** [comment]
Can i see the params for your AE model? And how long have you trained for?

**Reply by kapsner** [reply]
The AE-Model params were:

```
        self.autoencoder = AutoencoderKL(
            spatial_dims=3,
            in_channels=1,
            out_channels=1,
            num_res_blocks=2,
            num_channels=(32, 64, 64),
            attention_levels=(False, False, True),
            latent_channels=3,
            norm_num_groups=32,
            norm_eps=1e-6,
            with_encoder_nonlocal_attn=True,
            with_decoder_nonlocal_attn=True,
            use_flash_attention=False
        )
        self.discriminator = PatchDiscriminator(
            spatial_dims=3,
            num_channels=32,
            in_channels=1,
            out_channels= 1,
            num_layers_d=3,
            kernel_size=4,
            dropout=0.0
        )
```

The learning rate was 4.5e-06, batch-size 12, trained on 3xV100 for 1200 epochs; loss and reconstructions of the autoencoder showed pretty good results.

The LDM was trained for 2500 epochs with a lr=1e-6 and batch-size=64.

**Reply by marksgraham** [reply]
so a fair bit of training. what size is the latent space fed to the diffusion model then? generally the smaller the latent space the easier it is for the DDPM to model

**Reply by kapsner** [reply]
The latent-space size is `(3, 8, 24, 24)`; how do I get it smaller?

**Reply by marksgraham** [reply]
Another layer in your AE would reduce the size by a factor of 2 along each spatial dimension but it's already fairly small, I would expect your DDPM to work OK. In that tutorial the latent space is slightly smaller but it gets OK results in just a couple of hours of training on a single GPU.

probably something has been missed out in the conversion to lightning but I can't see anything obvious. have you tried loading your data into the LDM tutorial, which we know works, and seeing if that works for you?

**Reply by kapsner** [reply]
No, I haven't tried yet, but I'll give it a shot - I have also no idea to get along otherwise... I'll keep you updated on the results. Thanks already for your immediate help!!

**Reply by kapsner** [reply]
@marksgraham excuse me again, I have still doubts that I might be applying the sampling of the inferer falsely since the loss-curves of the training look like this: 
![grafik](https://github.com/Project-MONAI/GenerativeModels/assets/44973495/fe4cffc3-7b8f-4fbd-be51-1f8d86cdc28b)

So obviously, there is some learining progress (which looks similar to the progress in the [tutorial](https://github.com/Project-MONAI/GenerativeModels/blob/main/tutorials/generative/3d_ldm/3d_ldm_tutorial.ipynb)), but somehow, when calling `inferer.sample()`, the output seems as if there is no progress:

Synth-Example Step 6420:
![grafik](https://github.com/Project-MONAI/GenerativeModels/assets/44973495/cbcd0fd7-040b-4d4c-9dd2-1f0cd9b2feb8)


Synth-Example Step 49820:
![grafik](https://github.com/Project-MONAI/GenerativeModels/assets/44973495/82ca488c-472a-4329-a2a6-575fcd5d8004)

**Reply by marksgraham** [reply]
That is strange... I can't see anything obviously wrong with what you're doing with the inferer, though. It is possible we have someone broken it with some recent updates... What version of monai generative are you using? Have you tried just running our tutorial without changing anything (so on the data we provide in it) to check it gives sensible samples?

**Reply by kapsner** [reply]
It's version `0.2.2` from pip, so I guess there should nothing be related to recent updates.
I've not had time yet to get the tutorial running but I am trying now and will let you know the results!!

**Reply by kapsner** [reply]
So I've tried out the [tutorial](https://github.com/Project-MONAI/GenerativeModels/blob/main/tutorials/generative/3d_ldm/3d_ldm_tutorial.ipynb) and the sampled images also look strange (the only thing I changed was increasing the batch-size from 2 to 6):


![grafik](https://github.com/Project-MONAI/GenerativeModels/assets/44973495/54c885b2-2fdb-432e-a9c8-29d37fe542e7)

However, learning-curves look fine:
![grafik](https://github.com/Project-MONAI/GenerativeModels/assets/44973495/e1b56515-c40b-4daa-a37d-38ba5e5a3171)
![grafik](https://github.com/Project-MONAI/GenerativeModels/assets/44973495/56036c53-fdc9-4e44-83a6-4249bf4b5602)

**Reply by kapsner** [reply]
BTW I remember, that I once installed the latest main-branch from git over the pip-installed version, however I swiched back to pip later on, which should be the now installed version. Is there any option besides `python -c 'import generative; print(generative.__version__)'` to verify that I really use the pip-version

**Reply by marksgraham** [reply]
Not that i'm aware of - best to start on a clean env to check.

It is weird this isn't working. Could you try rolling back to an earlier version and seeing if it still does not work?

I'm away for the next week so I won't be able to look into this properly until the following week. let me know if you are still having problems then, especially if you can't get our tutorial to work, and I'll look into it properly on my end

**Reply by kapsner** [reply]
Hi @marksgraham ,

I've now tested the tutorial for versions `0.2.1` and `0.2.2` installed from pip, each time with a new environment and I get the same behaviour as described above (nice learning curves but trashy sampled synthetic images / only noise when sampling).

Environments were generated as follows:

```bash
conda create -n ldmdebug python=3.11
conda activate ldmdebug
conda install pytorch==2.1.0 torchvision==0.16.0 pytorch-cuda=11.8 -c pytorch -c nvidia
conda install ipykernel
pip install matplotlib nibabel tqdm lpips
pip install monai-generative==0.2.2 # ==0.2.1
```

**Reply by kapsner** [reply]
@marksgraham 

I've now checked the tutorial again with a batch-size = 2. I've also replaced the random noise generation with the original code, which I changed in order to make other batch-sizes work. Interestingly, this works:

```python
noise = torch.randn_like(z).to(device)
```

while replacing the noise-generation with this code produces the strange inferer-sampled images when training the ldm:

```python
input_size = ((len(images), ) + tuple(z.size())[1:])
noise = torch.rand(
    input_size,
    dtype=z.dtype,
    layout=z.layout,
    device=images.device
)
```

This, however, is necessary since the code crashes due to the batch-dimension of `z` not being equal to batch-size if the number of training data is not a multiple of the batch-size....

Do you have any clue whats wrong with the adapted noise-generation and why the code behaves like this?

**Reply by kapsner** [reply, marked as answer]
I found my mistake: I've accidentally used `torch.rand()`, which samples noise from a uniform distribution instead of `torch.randn()`, sampling from a normal distribution. So the batch-size-generic code to sample noise should look like this:

```python
input_size = ((len(images), ) + tuple(z.size())[1:])
noise = torch.randn(
    input_size,
    dtype=z.dtype,
    layout=z.layout,
    device=images.device
)
```

**Reply by kapsner** [reply]
After extensivley testing with our data, this works!

(A batch-size-generic noise-sampling can also be achieved this way: `noise = torch.randn_like(self.z_example[:len(images), ...]).to(images.device)`).

**Reply by marksgraham** [reply]
Glad it works.. the missing `n` was a subtle bug to spot!

**Comment by QuantPrincess** [comment]
@marksgraham @kapsner Hi. interestingly I ran into the same issue when trying to sample. I'm am fairly certain my autoencoding model of choice (vqgan) has converged and the size of the latent space is not that large (64,32,16,16). I even get the same type of noisy blobs when sampling the ddpm. I'm wondering if you did anything differently than just changing your noise distribution from rand to randn? If you have any other tips please let me know. else I just post my code. Really appreciate any tips you can give.
